data/lib/core/quests.lua

```diff
@@ -0,0 +1,113 @@
+Quest = {}
+Quest.__index = Quest
+
+function Quest:register()
+    local quests = Game.getQuests()
+    local missions = Game.getMissions()
+
+    self.id = #quests + 1
+
+    for _, mission in pairs(self.missions) do
+        mission.id = #missions + 1
+        mission.questId = self.id
+        missions[mission.id] = setmetatable(mission, Mission)
+    end
+
+    quests[self.id] = self
+    return true
+end
+
+function Quest:isStarted(player)
+    return player:getStorageValue(self.storageId) >= self.storageValue
+end
+
+function Quest:isCompleted(player)
+    for _, mission in pairs(self.missions) do
+        if not mission:isCompleted(player) then
+            return false
+        end
+    end
+    return true
+end
+
+function Quest:getMissions(player)
+    local missions = {}
+    for _, mission in pairs(self.missions) do
+        if mission:isStarted(player) then
+            missions[#missions + 1] = mission
+        end
+    end
+    return missions
+end
+
+function Quest:isTracking(key, value)
+    if self.storageId == key and value == self.storageValue then
+        return true
+    end
+
+    for _, mission in pairs(self.missions) do
+        if mission.storageId == key and value >= mission.startValue and value <= mission.endValue then
+            return true
+        end
+    end
+    return false
+end
+
+Mission = {}
+Mission.__index = Mission
+
+function Mission:isStarted(player)
+    local value = player:getStorageValue(self.storageId)
+    if value < self.startValue then
+        return false
+    end
+
+    if not self.ignoreEndValue and value > self.endValue then
+        return false
+    end
+
+    return true
+end
+
+function Mission:isCompleted(player)
+    if self.ignoreEndValue then
+        return player:getStorageValue(self.storageId) >= self.endValue
+    end
+    return player:getStorageValue(self.storageId) == self.endValue
+end
+
+function Mission:getName(player)
+    if self:isCompleted(player) then
+        return string.format("%s (Completed)", self.name)
+    end
+    return self.name
+end
+
+function Mission:getDescription(player)
+    if type(self.description) == "function" then
+        return self.description(player)
+    end
+
+    local value = player:getStorageValue(self.storageId)
+    if self.mainDescription then
+        local description = self.mainDescription:gsub("|STATE|", value)
+        description = self.mainDescription:gsub("\\n", "\n")
+        return description
+    end
+
+    if self.ignoreEndValue then
+        for current = self.endValue, self.startValue, -1 do
+            if value >= current then
+                return self.descriptions[current]
+            end
+        end
+    else
+        for current = self.endValue, self.startValue, -1 do
+            if value == current then
+                return self.descriptions[current]
+            end
+        end
+    end
```

Comment:
```suggestion
    if self.ignoreEndValue then
        for current = self.endValue, self.startValue, -1 do
            if value >= current and self.descriptions then
                return self.descriptions[current]
            end
        end
    else
        for current = self.endValue, self.startValue, -1 do
            if value == current and self.descriptions then
                return self.descriptions[current]
            end
        end
    end
```

this should get rid of this error
```
data/lib/core/quests.lua:107: attempt to index field 'descriptions' (a nil value)
stack traceback:
        [C]: in function '__index'
        data/lib/core/quests.lua:107: in function 'getDescription'
        data/lib/core/player.lua:752: in function 'sendQuestLine'
        ...cripts\eventcallbacks\player\default_onStorageUpdate.lua:43: in function 'callback'
        C:\ot\____SERVER\ots\data\scripts/lib\event_callbacks.lua:170: in function 'onQuestRequest'
        data/events/scripts/player.lua:658: in function <data/events/scripts/player.lua:656>
```

I'm not exactly sure how to reproduce it, here are my prompts
```
3:20 [0]: > Player("Player2"):setStorageValue(1003, 4)
03:20 [0]: > Player("Player2"):setStorageValue(1003, 3)
03:20 [0]: > Player("Player2"):setStorageValue(1003, 1)
03:20 [0]: > Player("Player2"):setStorageValue(1003, 2)
03:20 [0]: > Player("Player2"):setStorageValue(1002, 1)
```
and my config
```lua
Game.createQuest("Example Quest II", {
    storageId = 1002,
    storageValue = -1,

    missions = {
        {
            name = "Example Mission 1",
            storageId = 1002,
            startValue = -1,
            endValue = 10,
            ignoreEndValue = false,
            description = function(player)
                return string.format("Example description %d by %s.", player.storage[1003], player:getName())
            end
        },
        {
            name = "Example Mission 2",
            storageId = 1003,
            startValue = 1,
            endValue = 3,
            ignoreEndValue = false,
            description = "Mission 2"
        }
    }
}):register()
```

Reply:
I think -1 was never an accepted value, but it can be accepted by simply removing the "if value == -1 then" check in the onUpdateStorage event.

I also like the idea of ​​being able to set -1 by default to have quests that are always on and can be tracked from -1